src/actions/profile-view.js

```diff
@@ -104,12 +105,73 @@ export function changeMarkersSearchString(searchString: string): Action {
   };
 }
 
+/**
+ * This action is breaking the rules and using the getState function in the action
+ * creator. This is done because any stored CallNodePaths must be updated with the
+ * newly implementation filtered thread. Having this outside of the action creator
+ * and in components and tests makes it difficult to ensure that the CallNodePaths
+ * get properly updated, and would involve complex state management with hooks
+ * into the connected component's lifecycle.
+ */
 export function changeImplementationFilter(
-  implementation: ImplementationFilter
-): Action {
-  return {
-    type: 'CHANGE_IMPLEMENTATION_FILTER',
-    implementation,
+  nextImplementationFilter: ImplementationFilter
+): ThunkAction<void> {
+  return (dispatch, getState) => {
+    const previousImplementationFilter = URLStateSelectors.getImplementationFilter(
+      getState()
+    );
+    const threadIndex = URLStateSelectors.getSelectedThreadIndex(getState());
+    const { getFilteredThread } = ProfileViewSelectors.selectedThreadSelectors;
+
+    if (
+      previousImplementationFilter !== 'combined' &&
+      nextImplementationFilter !== 'combined'
+    ) {
+      // Going from 'cpp' to 'js' filtering means we need a fully reconstructed
+      // CallNodePath from an unfiltered thread, THEN go and filter to the other
+      // implementation.
```

Comment:
Maybe I miss something, but I don't see why we can't do this in the reducer ?

Reply:
I was thinking about this more and I may be able to. Let me take a look at it.